drivers/mspi/mspi_dw_vendor_specific.h

```diff
@@ -97,7 +97,226 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+
+/* Static allocation macros for vendor-specific data */
+#define VENDOR_SPECIFIC_DATA_DEFINE(inst) \
+	static QSPI_TRANSFER_LIST_Type mspi_dw_##inst##_transfer_list; \
+	static EVDMA_JOB_Type mspi_dw_##inst##_joblist[MAX_NUM_JOBS]; \
+	static const nordic_qspi_vendor_data_t mspi_dw_##inst##_vendor_data = { \
+		.transfer_list = &mspi_dw_##inst##_transfer_list, \
+		.joblist = &mspi_dw_##inst##_joblist[0] \
+	};
+
+#define VENDOR_SPECIFIC_DATA_GET(inst) &mspi_dw_##inst##_vendor_data
+
+/* Temporarily hard-coded as not in MDK yet */
+#define QSPI_TMOD_OFFSET	(0x490UL)
+#define QSPI_TMOD_TX_AND_RX	(0x0)
+#define QSPI_TMOD_TX_ONLY	(0x1)
+#define QSPI_TMOD_RX_ONLY	(0x2)
+static inline void vendor_specific_start_dma_xfer(const struct device *dev)
+{
+	struct mspi_dw_data *dev_data = dev->data;
+	const struct mspi_dw_config *config = dev->config;
+	const struct mspi_xfer_packet *packet =
+		&dev_data->xfer.packets[dev_data->packets_done];
+	NRF_QSPI_Type *preg = (NRF_QSPI_Type *)config->wrapper_regs;
+
+	/* Use vendor-specific data from config - stores job and transfer lists */
+	nordic_qspi_vendor_data_t *vendor_data = (nordic_qspi_vendor_data_t *)
+						 config->vendor_specific_data;
+
+	QSPI_TRANSFER_LIST_Type *transfer_list = vendor_data->transfer_list;
+	EVDMA_JOB_Type *joblist = vendor_data->joblist;
+
+	int tmod = 0;
+	int job_idx = 0;
+
+	/*
+	 * The Command and Address will always have a length of 4 from the DMA's
+	 * perspective. QSPI peripheral will use length of data specified in core registers
+	 */
+	if (dev_data->xfer.cmd_length > 0) {
+		joblist[job_idx++] = EVDMA_JOB(&packet->cmd, 4, EVDMA_ATTRIBUTE);
+	}
+	if (dev_data->xfer.addr_length > 0) {
+		joblist[job_idx++] = EVDMA_JOB(&packet->address, 4, EVDMA_ATTRIBUTE);
+	}
+
+	if (packet->dir == MSPI_TX) {
+		preg->CONFIG.RXTRANSFERLENGTH = 0;
+
+		if (packet->num_bytes > 0) {
+			joblist[job_idx++] = EVDMA_JOB(packet->data_buf, packet->num_bytes,
+						EVDMA_ATTRIBUTE);
+		}
+
+		/* Always terminate with null job */
+		joblist[job_idx] = EVDMA_NULL_JOB();
+		/* tx_job should point to first valid job */
+		transfer_list->tx_job = &joblist[0];
+
+		/* rx_job is always EVDMA_NULL_JOB() for transmit */
+		transfer_list->rx_job = &joblist[job_idx];
+		tmod = QSPI_TMOD_TX_ONLY;
+
+	} else {
+		preg->CONFIG.RXTRANSFERLENGTH = ((packet->num_bytes + dev_data->xfer.addr_length +
+						dev_data->xfer.cmd_length) >>
+						dev_data->bytes_per_frame_exp) - 1;
+
+		/* If sending address or command while being configured as controller */
+		if (job_idx > 0 && config->op_mode == MSPI_OP_MODE_CONTROLLER) {
+			tmod = QSPI_TMOD_TX_AND_RX;
+
+			/* After command and address, setup RX job for data */
+			joblist[job_idx++] = EVDMA_NULL_JOB();
+			joblist[job_idx++] = EVDMA_JOB(packet->data_buf, packet->num_bytes,
+						       EVDMA_ATTRIBUTE);
+			joblist[job_idx]   = EVDMA_NULL_JOB();
+
+			transfer_list->tx_job = &joblist[0];
+			transfer_list->rx_job = &joblist[job_idx-1];
```

Comment:
I think it would be clearer to have:
```
			transfer_list->rx_job = &joblist[job_idx];
			joblist[job_idx++] = EVDMA_JOB(packet->data_buf, packet->num_bytes,
						       EVDMA_ATTRIBUTE);
```
Similarly for TX jobs - `transfer_list->tx_job = &joblist[job_idx];` could be put at the very beginning of the common path.